agent/utils/utils.go

```diff
@@ -140,29 +136,73 @@ func hasVMServiceAccount() bool {
 	return true
 }
 
-func checkVMID() {
-	// VM Identity requires a service account.
-	hasVMID = hasVMServiceAccount()
+func getVMID(audience string) string {
+	for {
+		idPath := fmt.Sprintf("instance/service-accounts/default/identity?format=full&audience=%s", audience)
+		vmID, err := metadata.Get(idPath)
+		if err == nil {
+			return vmID
+		}
+		log.Printf("failure fetching a VM ID: %v", err)
+	}
+}
+
+func newIDChannel(ctx context.Context, audience string) chan string {
+	ch := make(chan string)
+	go func() {
+		defer close(ch)
+		ticker := time.NewTicker(10 * time.Second)
+		defer ticker.Stop()
+
+		currID := getVMID(audience)
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-ticker.C:
+				currID = getVMID(audience)
+			case ch <- currID:
+				continue
+			}
+		}
+	}()
+	return ch
+}
+
+type vmTransport struct {
+	proxyURL string
+	wrapped  http.RoundTripper
+	idChan   chan string
 }
 
-// addVMIDHeader adds a header to the given request identifying the VM (if any) on which the agent is running.
+func (t *vmTransport) RoundTrip(r *http.Request) (*http.Response, error) {
+	select {
+	case <-r.Context().Done():
+		return nil, fmt.Errorf("request cancelled")
+	case id := <-t.idChan:
```

Comment:
This means, if the fe client request rate is higher than the ticker rate ( which is currently set to 10 seconds), the client would experience up to 10 seconds delays.

I would use a simple variable for caching here and not a channel. You can wait on the initial value and use a way higher refresh rate. The token expires in order of multiple minutes as far as I recall.

Reply:
> This means, if the fe client request rate is higher than the ticker rate ( which is currently set to 10 seconds), the client would experience up to 10 seconds delays.

No it doesn't.

The idChan here is constantly filled up regardless of whether or not the ticker ticks. The 10 second ticker rate just affects how frequently the value is updated, not how frequently it is sent on the channel.

That being said, I will try out the approach of caching a value (along with the required locking) and see if I like that better.